### PR TITLE
Step8 Add System Tests for Task CRUD Operations

### PR DESCRIPTION
**Pull request by fanfannb**
Background
As part of our efforts to ensure robustness in our task management system, this PR introduces comprehensive system tests covering the basic CRUD operations for tasks. These tests will help us maintain confidence in our application's functionality as we continue to develop and refactor.

Changes
Added system tests for creating, viewing, editing, and deleting tasks.
Configured Capybara to use Selenium.

Tests
The following scenarios are covered:

Creating a Task: Tests if a user can create a task with valid inputs and verifies the success message and the presence of the task details on the page.
Viewing a Task: Tests if the task details page shows the correct task information.
Editing a Task: Tests if a user can edit a task and confirm the changes are reflected properly.
Deleting a Task: Tests if a task can be deleted with a confirmation alert and ensures the task is no longer visible on the list page.
Each test is isolated. 

docker-compose run api rails spec
截屏2024-05-29 16 37 41

### ステップ8: テスト(system spec)を書こう
- First make sure that these gems are exist in Gemfile
  ```
  group :test do
    gem 'capybara', '>= 2.15'
    gem 'selenium-webdriver'
  end
  ```
  **Note**: Remove the `webdrivers` gem from your Gemfile. If `webdrivers` is present, it will attempt to  find Chrome in your application’s container. 
  As Chrome isn’t installed  in the Dockerfile, the spec will fail.
  
- Before start testing we need to register a new driver with Capybara that is configured to use the Selenium container, add the below codes to 
  `spec/rails_helper.rb`
  ```
  Capybara.register_driver :remote_chrome do |app|
    hub_url = 'https://chrome:4444/wd/hub'
    chrome_capabilities = ::Selenium::WebDriver::Remote::Capabilities.chrome(
      'goog:chromeOptions' => {
        'args' => %w[no-sandbox headless disable-gpu window-size=1680,1050],
      },
    )
    Capybara::Selenium::Driver.new(app, browser: :remote, url: hub_url, desired_capabilities: chrome_capabilities)
  end

  config.before(:each, type: :system) do
    driven_by :rack_test
  end

  config.before(:each, type: :system, js: true) do
    driven_by :remote_chrome
    Capybara.server_host = IPSocket.getaddress(Socket.gethostname)
    Capybara.server_port = 3000
    Capybara.app_host = "https://#{Capybara.server_host}:#{Capybara.server_port}"
  end
  ```
- specを書くための準備をしましょう
  - `spec/spec_helper.rb` 、 `spec/rails_helper.rb` を用意しましょう
- ~~feature spec~~ system specをタスク機能に対して書きましょう
  - Rails 5.1 以降、新たにsystem testの機能を追加しました
    - [日本語](https://qiita.com/jnchito/items/c7e6e7abf83598a6516d), [英語](https://rossta.net/blog/why-rails-system-tests-matter.html)
  - dockerを利用して研修を行う場合、以下の設定が必要です。
    1. [Dockerfile](https://qiita.com/ngron/items/f61b8635b4d67f666d75#failed-to-read-the-sessionstorage-property-from-window-storage-is-disabled-inside-data-urls)
    2. [spec/rails_helper.rb](https://commis.hatenablog.com/entry/2018/11/16/171608)
  - feature specですと `database_cleaner` という gemは必要でしたが、 system specに変更することで `database_cleaner` の導入が要らなくなった
- Circle CIなどのCIツールを導入して、Slackに通知するようにしましょう
  - Fablic/training内でPRのやり取りをする場合、CIツールの導入は任意(optional)です。CircleCIのAdmin権限が無いので、`.circleci/config.yml`を設定しても実行できないです。
- 参考書籍：https://leanpub.com/everydayrailsrspec-jp

### CODE REVIEW

File: myapp/spec/system/tasks_spec.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe 'Task management', type: :system do
```
Comment (kanipan) [inline myapp/spec/system/tasks_spec.rb]: こう言う感じで分けるとファイルの中身を見なくてもどのテストがあるのかが分かりやすくなります。

```
myapp/spec/system/tasks/new_spec.rb
myapp/spec/system/tasks/show_spec.rb
myapp/spec/system/tasks/edit_spec.rb
myapp/spec/system/tasks/delete_spec.rb
```

describe はより具体的に書くとより良いです。

```
RSpec.describe 'visit /task/new', type: :system do
```

**Review by kanipan — APPROVED** [review]

**Comment by kanipan** [comment]
LGTM!

optional https://github.com/Fablic/training/pull/1652/files#r1619789930

寿司ゆき:smile

**Review by mvincent-yap — APPROVED** [review]
LGTM

**Review by kanipan — APPROVED** [review]

**Comment by kanipan** [comment]
追い LGTM

寿司ゆき:good_night

**Pull request merged**